synapse/handlers/message.py

```diff
@@ -1382,13 +1382,14 @@ async def persist_and_notify_client_event(
 
         if event.type == EventTypes.Member:
             if event.content["membership"] == Membership.INVITE:
-                event.unsigned[
-                    "invite_room_state"
-                ] = await self.store.get_stripped_room_state_from_event_context(
-                    context,
-                    self.room_prejoin_state_types,
-                    membership_user_id=event.sender,
-                )
+                if not event.internal_metadata.outlier:
```

Comment:
This seems like it would be a problem with any outlier that includes `invite_room_state`. With MSC2716, we create a lot of outliers though which makes sense why it's obvious to encounter.

Reply:
I'm a bit confused why this is only a problem when using workers? The same code paths should be hit either way?

Reply:
@erikjohnston Agreed, it seems that way to me as well ("This seems like it would be an existing problem with any `outlier` that includes `invite_room_state`"). I haven't reproduced this myself.

What's a good way to test worker mode if it is that specific?

@tulir Can you share some reproduction steps to lead us here?

Reply:
Just batch sending any invite event in `state_events_at_start` seems to reproduce it. Sending the same invite event without workers doesn't reproduce it.

It looks like on the master process, `current_state_ids` will be an empty dict, but on the worker it will be `None`.

Reply:
It looks like the [Synapse Complement test runner](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/scripts-dev/complement.sh#L41-L54) does have the option to have the `WORKERS` environment variable set and go into worker mode. But I unable to get it working,

```
WORKERS=1 COMPLEMENT_ALWAYS_PRINT_SERVER_LOGS=1 COMPLEMENT_DIR=../complement ./scripts-dev/complement.sh TestImportHistoricalMessages
```

Results in my Docker crashing in different ways:

```
Fatal Error

Docker daemon failed to start

2021-11-10T00:39:12Z dockerd time="2021-11-10T00:39:11.027705833Z" level=info msg="[graphdriver] using prior storage driver: overlay2"
```

```
2021-11-10T03:24:22Z dockerd time="2021-11-10T03:24:22.696959125Z" level=info msg="Container 2faad2c3366d177fba0d1df67959518ac66114e72ddca539434f6c53cdadc084 failed to exit within 10 seconds of signal 15 - using the force
```

---

> It looks like on the master process, `current_state_ids` will be an empty dict, but on the worker it will be `None`.

Could we additionally fix up the difference in behavior?

Reply:
I didn't figure out where the difference is coming from yet since the `current_state_ids` is loaded somehow weirdly

Reply:
It looks like `_current_state_ids` is initialized to an [empty dict for outlier events](https://github.com/matrix-org/synapse/blob/v1.47.0rc2/synapse/events/snapshot.py#L134-L140), but that field isn't preserved when the `EventContext` is serialized/deserialized

Reply:
Made a change to include outlier metadata in serialized event contexts and reverted the other one, seems to work still